File: app/src/main/java/com/johnpetitto/orachat/data/ResponseTransformer.java
package com.johnpetitto.orachat.data;

import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.SingleTransformer;
import io.reactivex.annotations.NonNull;

public class ResponseTransformer<T> implements SingleTransformer<ApiResponse<T>, T> {
    @Override
    public SingleSource<T> apply(@NonNull Single<ApiResponse<T>> upstream) {
        return upstream.flatMap(response -> {
            String errorMessage = response.getMessage();
            if (errorMessage != null) {
                return Single.error(new Throwable(errorMessage));
            } else {
                return Single.just(response.getData());
            }
        });
    }
}

File: app/src/main/java/com/johnpetitto/orachat/data/ResponseTransformer.kt
package com.johnpetitto.orachat.data

import io.reactivex.Single
import io.reactivex.SingleSource
import io.reactivex.SingleTransformer

class ResponseTransformer<T>: SingleTransformer<ApiResponse<T>, T> {
    override fun apply(upstream: Single<ApiResponse<T>>): SingleSource<T> =
            upstream.flatMap { (data, errorMessage) ->
                if (data != null) {
                    Single.just(data)
                } else {
                    Single.error(Throwable(errorMessage))
                }
            }
}
